superset/connectors/sqla/api.py

```diff
@@ -0,0 +1,177 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import logging
+from typing import Dict, List
+
+from flask import current_app, g
+from flask_appbuilder.models.sqla.interface import SQLAInterface
+from flask_babel import lazy_gettext as _
+from marshmallow import fields, post_load, validates_schema, ValidationError
+from marshmallow.validate import Length
+from sqlalchemy.exc import SQLAlchemyError
+
+from superset.connectors.sqla.models import SqlaTable
+from superset.constants import RouteMethod
+from superset.models.core import Database
+from superset.views.base import DatasourceFilter, get_datasource_exist_error_msg
+from superset.views.base_api import BaseOwnedModelRestApi
+from superset.views.base_schemas import BaseOwnedSchema, validate_owner
+
+logger = logging.getLogger(__name__)
+
+
+def validate_database(value):
+    item = (
+        current_app.appbuilder.get_session.query(Database)
+        .filter_by(id=value)
+        .one_or_none()
+    )
+    if not item:
+        g.tmp_database = None
+        raise ValidationError(_("Database does not exist"))
+    # Database exists save it on g to save further db round trips
+    g.tmp_database = item
+
+
+def validate_table_exists(data: Dict):
+    if "table_name" not in data:
+        return
+    table_name: str = data["table_name"]
+    try:
+        if g.tmp_database:
+            g.tmp_database.get_table(table_name, schema=data.get("schema", ""))
+    except SQLAlchemyError as e:
+        logger.exception(f"Got an error {e} validating table: {table_name}")
+        raise ValidationError(
+            _(
+                f"Table [{table_name}] could not be found, "
+                "please double check your "
+                "database connection, schema, and "
+                f"table name, error: {e}"
+            )
+        )
+
+
+def validate_table_uniqueness(data: Dict):
+    if not ("database" in data and "table_name" in data):
+        return
+    database_name: str = data["database"]
+    table_name: str = data["table_name"]
+
+    with current_app.appbuilder.get_session.no_autoflush:
+        table_query = current_app.appbuilder.get_session.query(SqlaTable).filter(
+            SqlaTable.table_name == table_name, SqlaTable.database_id == database_name
+        )
+        if current_app.appbuilder.get_session.query(table_query.exists()).scalar():
+            raise ValidationError(get_datasource_exist_error_msg(table_name))
+
+
+class TablePostSchema(BaseOwnedSchema):
+    __class_model__ = SqlaTable
+
+    database = fields.Integer(validate=validate_database)
+    schema = fields.String()
+    table_name = fields.String(required=True, validate=Length(1, 250))
+    owners = fields.List(fields.Integer(validate=validate_owner))
+
+    @validates_schema
+    def validate_schema(self, data: Dict):  # pylint: disable=no-self-use
+        validate_table_uniqueness(data)
```

Comment:
Mixing API schema validation and business logic/db validation feels like a bad pattern here. Marshmallow should be concerned with keeping requests in/out of the API consistent, whereas data access rules and/or auth should live inside the ViewController.

One other interesting side note - these checks could potentially trigger BEFORE auth validation happens, which could lead to leakage (a user without rights to a DB for instance could phish around for DB names which would trigger these replies if auth logic happened downstream of this)

Reply:
Agreed, good catch.

Reply:
The schema validation is debatable, since we are only performing basic data validation (does the relation exist?, is the fields unique?). Yet I have mixed feeling about it. The other approach would use `pre_`, `post_` hooks. Note that using marshmallow validation gives us proper error messages  for free.

It's not possible for the validation to trigger before any auth validation. Since this is done at the viewcontrol decorator level.

Reply:
I don't think it's one or the other. I think marshmallow validation is great for validating the shape of the request. I do not think that marshmallow should be running database queries. My recommendation would be for the validation/auth logic to look something like:

1. Can the user access the route? If yes, continue, if no 403
2. Run marshmallow validation to validate the shape of the request. If OK, continue, if not 400
3. Look up the resource being modified. Can the user modify it? If so, continue, if not 403
4. Attempt the resource modification.

I think that can be accomplished via some combination of marshmallow and pre_ and post_ hooks, but you're the expert :)

Reply:
This can also become a security issue quite easily. If we are doing data validation inside of marshmallow schemas, an attacker could use validation errors to gather information about a resource that they should not have access to.